### PR TITLE
OGL: Use PBO for framedump, with async readback.

### PR DESCRIPTION
**Pull request by degasus**
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4344)

### CODE REVIEW

**Comment by lioncash** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1582 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpRyKhCniQiBL9wpGP:-KTpRyKhCniQiBL9wpGQ:b-82t30i) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.cpp#L1582)):*
> ```cpp
> void Renderer::FlushFrameDump()
> {
>   if (m_last_frame_exported)
> ```

```cpp
if (!m_last_frame_exported)
  return;
```

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1598 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpS2Ie0krODrJIXFOw:-KTpS2Ie0krODrJIXFOx:b6u5ed6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.cpp#L1598)):*
> ```cpp
> void Renderer::DumpFrame(const TargetRectangle& flipped_trc, u64 ticks)
> {
>   if (IsFrameDumping())
> ```

```cpp
if (!IsFrameDumping())
  return;
```

---

*[Source/Core/VideoBackends/OGL/Render.h, line 119 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpS6q139T11N-hUUyp:-KTpS6q139T11N-hUUyq:b-1hj32p) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.h#L119)):*
> ```objc
> 
>   // avi dumping state to delay one frame
>   u32 m_frame_dumping_pbo[2] = {};
> ```

`std::array`

---

*[Source/Core/VideoBackends/OGL/Render.h, line 120 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpS8wSeKfvRw7q_iht:-KTpS8wT-bMK10RdMB0u:b-1hj32p) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.h#L120)):*
> ```objc
>   // avi dumping state to delay one frame
>   u32 m_frame_dumping_pbo[2] = {};
>   bool m_frame_pbo_is_mapped[2] = {};
> ```

`std::array`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344)*

**Comment by degasus** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1582 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpRyKhCniQiBL9wpGP:-KVswcEGvVVEi1hjr5Gg:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.cpp#L1582)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> if (!m_last_frame_exported)
>   return;
> ```
</details>
Done.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1598 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpS2Ie0krODrJIXFOw:-KVswiM3marwHDyai4P_:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.cpp#L1598)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> if (!IsFrameDumping())
>   return;
> ```
</details>
Done.

---

*[Source/Core/VideoBackends/OGL/Render.h, line 119 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpS6q139T11N-hUUyp:-KVswuLJTt9rbFr7DQ-O:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.h#L119)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `std::array`
</details>
Done.

---

*[Source/Core/VideoBackends/OGL/Render.h, line 120 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KTpS8wSeKfvRw7q_iht:-KVswukFaTk5ZrZVNc_d:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6ef486431600b0db38304bbbd70b243ea4542947/Source/Core/VideoBackends/OGL/Render.h#L120)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `std::array`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344)*

**Comment by stenzek** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1626 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KVtfB51nHVW5Sn9kwV1:-KVtfB51nHVW5Sn9kwV2:bs1upgg) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4f9b4acf2c9168ed4b62588b47f79b8240418fa0/Source/Core/VideoBackends/OGL/Render.cpp#L1626)):*
> ```cpp
> 
>   glBufferData(GL_PIXEL_PACK_BUFFER, m_last_frame_width * m_last_frame_height * 4, nullptr,
>                GL_STREAM_READ);
> ```

Would it be worth storing the dimensions of the buffer here to avoid the glBufferData() for each frame? Not sure if drivers are smart enough to avoid reallocating the buffer each time even if the size is the same.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1627 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KVtfbkurQ_3uE6HPNgl:-KVtfbkurQ_3uE6HPNgm:bib4lam) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4f9b4acf2c9168ed4b62588b47f79b8240418fa0/Source/Core/VideoBackends/OGL/Render.cpp#L1627)):*
> ```cpp
>   glBufferData(GL_PIXEL_PACK_BUFFER, m_last_frame_width * m_last_frame_height * 4, nullptr,
>                GL_STREAM_READ);
>   glPixelStorei(GL_PACK_ALIGNMENT, 1);
> ```

Would 4 be more appropriate here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344)*

**Comment by RisingFog** [comment]
The savestate issues are fixed with the recent PR merged, functionally seems to be working as it should.

:lgtm: 

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-:-KVuQyvlJo_M1eT1axZ7:bqivaje)*

**Comment by degasus** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1626 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KVtfB51nHVW5Sn9kwV1:-KVxhhD0pCPaVgKdW0IJ:b-88c0ju) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4f9b4acf2c9168ed4b62588b47f79b8240418fa0/Source/Core/VideoBackends/OGL/Render.cpp#L1626)):*
<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> Would it be worth storing the dimensions of the buffer here to avoid the glBufferData() for each frame? Not sure if drivers are smart enough to avoid reallocating the buffer each time even if the size is the same.
</details>
The driver usually optimize this case, but doing it ourself is likely faster. Not sure if it's worth it through ;)
I'll update the PR for it.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1627 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KVtfbkurQ_3uE6HPNgl:-KVxiDSTAjU57x0oVjXQ:b-2r5cqy) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4f9b4acf2c9168ed4b62588b47f79b8240418fa0/Source/Core/VideoBackends/OGL/Render.cpp#L1627)):*
<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> Would 4 be more appropriate here?
</details>
As we use RGBA, this has no effect. But I can change it if you want. Bad luck, there is no way to query the best value here, as 16 is likely faster, but not always allowed by the GL specs...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344)*

**Comment by degasus** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1626 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KVtfB51nHVW5Sn9kwV1:-KW-h2E2P_S5TU-od3wN:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4f9b4acf2c9168ed4b62588b47f79b8240418fa0/Source/Core/VideoBackends/OGL/Render.cpp#L1626)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> The driver usually optimize this case, but doing it ourself is likely faster. Not sure if it's worth it through ;)
> I'll update the PR for it.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344)*

**Comment by stenzek** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/VideoBackends/OGL/Render.cpp, line 1626 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344#-KVtfB51nHVW5Sn9kwV1:-KW2HDb7bRNRGtrVriNB:b-rivr4j) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4f9b4acf2c9168ed4b62588b47f79b8240418fa0/Source/Core/VideoBackends/OGL/Render.cpp#L1626)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Done.
</details>
Sounds good to me. If the bloat is a concern you could throw all the variables into a struct instead, and swap that, but I'm fine with it either way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4344)*

File: Source/Core/VideoBackends/OGL/Render.cpp [hunk 1]
```diff
@@ -767,6 +767,9 @@ Renderer::Renderer()
 
 Renderer::~Renderer()
 {
+  FlushFrameDump();
+  if (m_frame_dumping_pbo[0])
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/Render.cpp]: meh, missed one call here:
FinishFrameData();

**Pull request merged**